src/Persistence/Sql.php

```diff
@@ -945,21 +952,34 @@ public function getFieldSqlExpression(Field $field, Expression $expression)
         return $expression->expr($mask, $prop);
     }
 
-    /**
-     * Last ID inserted.
-     *
-     * @return mixed
-     */
-    public function lastInsertId(Model $model)
+    private function getIdSequenceName(Model $model): ?string
     {
-        $seq = $model->sequence ?: null;
+        $sequenceName = $model->sequence ?: null;
 
-        // PostgreSQL PDO always requires sequence name in lastInsertId method as parameter
-        // So let's use its default one if no specific is set
-        if ($this->connection instanceof \atk4\dsql\Postgresql\Connection && $seq === null) {
-            $seq = $model->table . '_' . $model->id_field . '_seq';
+        if ($sequenceName === null) {
+            // PostgreSQL uses sequence internally for PK autoincrement,
+            // use default name if not set explicitly
+            if ($this->connection instanceof \atk4\dsql\Postgresql\Connection) {
+                $sequenceName = $model->table . '_' . $model->id_field . '_seq';
+            }
         }
 
-        return $this->connection->lastInsertId($seq);
+        return $sequenceName;
+    }
+
+    public function lastInsertId(Model $model): string
+    {
+        return $this->connection->lastInsertId($this->getIdSequenceName($model));
+    }
+
+    protected function syncIdSequence(Model $model): void
+    {
+        // PostgreSQL sequence must be manually synchronized if a row with explicit ID was inserted
+        if ($this->connection instanceof \atk4\dsql\Postgresql\Connection) {
+            $this->connection->expr(
+                'select setval([], coalesce(max({}), 0) + 1, false) from {}',
+                [$this->getIdSequenceName($model), $model->id_field, $model->table]
```

Comment:
This could be quite inefficient to select `max(id)` on each insert if table is very big. Isn't it?

Also not sure how this helps if called in case I already set id=999 manually while inserting.
For example, I have:
* in table records 1,2,3
* sequence currval is at 2
* then insert new record and also set specific id for it id=9;

Then if i'm not mistaken, it will also call this method and set sequence currval to `3+1=4` and later on will insert record with id=9. And at that point sequence currval is not consistent with inserted data (currval after insert should be 9).

Reply:
really? ID SHOULD have always index (and ALWAYS implied if ID is PK (=unique index))

Reply:
if you insert 9, then next ID must be 10, where it the issue?

Reply:
Yes next ID will be fine, but sequence.currval *after* inserting 9 but before inserting next record will be wrong = 4.